components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/EngineDelegateMiddleware.kt

```diff
@@ -138,8 +142,13 @@ internal class EngineDelegateMiddleware(
         store: Store<BrowserState, BrowserAction>,
         action: EngineAction.SaveToPdfAction,
     ) = scope.launch {
-        getEngineSessionOrDispatch(store, action)
-            ?.requestPdfToDownload()
+        try {
+            getEngineSessionOrDispatch(store, action)
+                ?.requestPdfToDownload()
+        } catch (e: IOException) {
```

Comment:
Where is the `IOException` thrown? I only see an `IllegalStateException` thrown [here](https://searchfox.org/mozilla-central/rev/12a18f7e112a4dcf88d8441d439b84144bfbe9a3/widget/android/nsWindow.cpp#1779).

In my testing, this catch block is not invoked.

Reply:
In fact, when changing catch to just `Exception`, I still do not see this catch being triggered.

I do see [this log](https://searchfox.org/mozilla-central/rev/12a18f7e112a4dcf88d8441d439b84144bfbe9a3/widget/android/nsWindow.cpp#1780) showing up in my logs though:

```
GeckoEngineSession      org.mozilla.fenix.debug              E  Save to PDF failed.
                                                                java.lang.IllegalStateException: Coud not save this page as PDF.
                                                                	at org.mozilla.gecko.mozglue.GeckoLoader.nativeRun(Native Method)
                                                                	at org.mozilla.gecko.GeckoThread.run(GeckoThread.java:677)
```

The clue here is that the logger that is reporting this is `GeckoEngineSession`.  Which means that our `GeckoResult.exceptionListener` block is safely catching this exception (which is good!).

What this means for us is that we do not need to annotate or wrap `GeckoSession.saveAsPdf()` as a call that can throw. We need instead `notifyObservers` in that exception listener block about the error and dispatch an action accordingly.

cc: @ohall-m for reference since we were talking about this a few weeks ago as well over [here](https://github.com/mozilla-mobile/android-components/issues/12642#issuecomment-1276808690).

Reply:
Could you help me to clarify, should I close this PR then?

Reply:
A little more info that I have - The `IllegalStateException` is the [first exception](https://searchfox.org/mozilla-central/rev/2d1097c27f31cb15f7588b043d691cda9710d14b/widget/android/nsWindow.cpp#1779,1789,1803) that could happen. The `GeckoResult` will complete exceptionally with this exception. This would be early in the process.

The `IOException` could happen after the `GeckoResult` completes and the caller starts reading the stream. ([Exception](https://searchfox.org/mozilla-central/rev/ffa4d00965c5281def6d3ddcbcdf6259d38c9b9a/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoInputStream.java#129,161,166-167) only happens on read/write). This error would be very late in the process.

Reply:
@ohall-m Ah okay, so there isn't any reasonable way to test the code path for the `IOException` if I wanted to see the flow?

That said, I think we still need to handle the `IllegalStateException` and propagate that error in the same manner - I was able to organically run into that quite easily by going to 'Save to PDF' before the first paint to the screen was completed.

Reply:
Not entirely sure how we could test the`IOException`, but maybe something like navigating away or closing the tab while a large page is processing the save to PDF? Think the timing would be very narrow to after the `GeckoResult` completes, but before Gecko finishes writing. 

(Something like [1](https://searchfox.org/mozilla-central/rev/968bd894205cf4f579d94ac4e175cc3187458605/docshell/base/CanonicalBrowsingContext.cpp#734) or [2](https://searchfox.org/mozilla-central/rev/968bd894205cf4f579d94ac4e175cc3187458605/dom/base/nsGlobalWindowOuter.cpp#5134,5147,5154,5166,5176,5181,5217,5227,5259,5266) has to go wrong - any of those should propagate back up as an `IOException` because the file didn't complete correctly.)

Reply:
> the caller starts reading the stream.

Arturo and I spoke offline and he mentioned this part that you wrote out which I missed, and that he found a good fix for: we already have the code in Downloads that [handles this](https://github.com/mozilla-mobile/android-components/blob/1b2d782303ba8eb4dc4778d95c69d725d882d864/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L774-L782) so we don't need this newly added try-catch block.

While we don't have _specific_ code to handle the PDF IOException, the Downloads code path already tests and covers this use case. Sorry for the confusion!